src/lightning_app/runners/cloud.py

```diff
@@ -243,367 +137,622 @@ def dispatch(
+        self,
+        cluster_id: str,
+        project_id: str,
+        cloudspace_name: str,
+        cloudspace_id: str,
+        run_id: str,
+        desired_state: V1LightningappInstanceState,
+        queue_server_type: V1QueueServerType,
+        env_vars: List[V1EnvVar],
+        auth: V1LightningAuth,
+    ) -> Externalv1LightningappInstance:
+        """Create a new instance of the given run with the given specification."""
+        return self.backend.client.cloud_space_service_create_lightning_run_instance(
             project_id=project_id,
-            body=V1ProjectClusterBinding(cluster_id=cluster_id, project_id=project_id),
+            cloudspace_id=cloudspace_id,
+            id=run_id,
+            body=IdGetBody1(
+                cluster_id=cluster_id,
+                name=cloudspace_name,
+                desired_state=desired_state,
+                queue_server_type=queue_server_type,
+                env=env_vars,
+                auth=auth,
+            ),
         )
 
-    def _get_default_cluster(self, project_id: str) -> str:
-        """This utility implements a minimal version of the cluster selection logic used in the cloud.
+    @staticmethod
+    def _api_package_and_upload_repo(repo: LocalSourceCodeDir, run: V1LightningRun) -> None:
+        """Package and upload the provided local source code directory to the provided run."""
+        if run.source_upload_url == "":
+            raise RuntimeError("The source upload url is empty.")
+        repo.package()
+        repo.upload(url=run.source_upload_url)
 
-        TODO: This should be requested directly from the platform.
-        """
-        cluster_bindings = self.backend.client.projects_service_list_project_cluster_bindings(
-            project_id=project_id
-        ).clusters
+    @staticmethod
+    def _print_specs(run_body: CloudspaceIdRunsBody, print_format: str) -> None:
+        """Print the given run body in either `web` or `gallery` format."""
+        if print_format not in ("web", "gallery"):
+            raise ValueError(
+                f"`LIGHTNING_CLOUD_PRINT_SPECS` should be either `web` or `gallery`. You provided: {print_format}"
+            )
 
-        if not cluster_bindings:
-            raise ValueError(f"No clusters are bound to the project {project_id}.")
+        flow_servers_json = [{"Name": flow_server.name} for flow_server in run_body.flow_servers]
+        logger.info(f"flow_servers: {flow_servers_json}")
+        works_json = json.dumps(_to_clean_dict(run_body.works, print_format == "web"), separators=(",", ":"))
+        logger.info(f"works: {works_json}")
+        logger.info(f"entrypoint_file: {run_body.app_entrypoint_file}")
+        requirements_path = getattr(getattr(run_body.image_spec, "dependency_file_info", ""), "path", "")
+        logger.info(f"requirements_path: {requirements_path}")
 
-        if len(cluster_bindings) == 1:
-            return cluster_bindings[0].cluster_id
+    def dispatch(
```

Comment:
I suggest putting the public methods at the top of the class

Reply:
Done 😃